main/webapp/modules/core/scripts/dialogs/expression-preview-dialog.js

```diff
@@ -127,14 +127,53 @@ ExpressionPreviewDialog.Widget = function(
     });
         
     var self = this;
+    
+    this._expressionHistory = [];
+    this._currentHistoryIndex = -1; // -1 is current expression, 0 is most recent history, 1 is next history
+
     this._elmts.expressionPreviewTextarea
         .val(this.expression)
         .on("keyup change input",function(){
+            self._expressionHistory[self._currentHistoryIndex] = self._getExpression();
             self._scheduleUpdate();
         })
         .trigger('select')
         .trigger('focus');
 
+    this._elmts.expressionPreviewTextarea
+        .on("keydown", function(e) {
+            var setExpression = function(exp) {
+                var o = Scripting.parse(exp);
+                self._elmts.expressionPreviewTextarea[0].value = o.expression;
+                self._elmts.expressionPreviewLanguageSelect[0].value = o.language;
+                
+                self.update();
+            }
+
+            // handle up and down arrow keys
+            if (e.keyCode == 38) {
```

Comment:
It looks like using `keyCode` [is deprecated](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardEvent/keyCode) (and is also less clear than using `key`). I think the conditional here and on line 165 could be `e.key === "ArrowUp"` and `e.key === "ArrowDown"`, respectively.